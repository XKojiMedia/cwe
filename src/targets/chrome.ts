import fs from 'fs';
import path from 'path';
import { CweConfig, ManifestIcons, ExtensionInfo, ExtensionCompilerOption, ExtensionCompiler } from '../config';
import archiver from 'archiver';
import { ensureFile } from 'fs-extra';

interface ChromeExtensionBrowserAction {
  default_icon?: ManifestIcons;
}
interface ChromeExtensionBackground {
  scripts: string[];
}
interface ChromeExtensionOptionsUi {
  page: string;
  open_in_tab: boolean;
}

interface ChromeExtension {
  manifest_version: 2;
  name: string;
  short_name: string;
  description: string;
  version: string;
  icons: ManifestIcons;
  browser_action: ChromeExtensionBrowserAction;
  permissions: string[];
  content_security_policy: string;
  background: ChromeExtensionBackground;
  options_ui?: ChromeExtensionOptionsUi;
  offline_enabled: boolean;
}

const buildBrowserAction = (config: CweConfig): ChromeExtensionBrowserAction => {
  if (config.manifestOptions?.browserAction) {
    return {
      default_icon: config.manifestOptions.browserAction.defaultIcon,
    };
  }
  return {};
};

const buildOptionsUi = (config: CweConfig): ChromeExtensionOptionsUi | undefined => {
  if (config.manifestOptions?.settingsOptions) {
    return {
      page: config.manifestOptions.settingsOptions.page,
      open_in_tab: config.manifestOptions.settingsOptions.openInTab,
    };
  }
  return undefined;
};

const buildExtensionData = (config: CweConfig): ChromeExtension | null => {
  if (!config.manifestOptions) {
    return null;
  }

  return {
    manifest_version: 2,
    name: config.manifestOptions.name,
    short_name: config.manifestOptions.shortName,
    description: config.manifestOptions.description,
    version: config.manifestOptions.version,
    icons: config.manifestOptions.icons,
    browser_action: buildBrowserAction(config),
    permissions: config.manifestOptions.permissions,
    content_security_policy: config.manifestOptions.contentSecurityPolicy,
    background: config.manifestOptions.backgroundOptions,
    options_ui: buildOptionsUi(config),
    offline_enabled: config.manifestOptions.offlineEnabled,
  };
};

export const generateExtensionInfo = async(config: CweConfig): Promise<ExtensionInfo> => {
  return {
    content: JSON.stringify(buildExtensionData(config), null, 2),
    fileName: 'manifest.json',
    fileType: 'json',
  };
};

export const compileExtension: ExtensionCompiler = async(opts: ExtensionCompilerOption) => {
  const outputPath = path.resolve(opts.extensionBuildOutputDir, 'chrome.zip');
  await ensureFile(outputPath);
  const output = fs.createWriteStream(outputPath);

  return new Promise((resolve, reject) => {

    // zip extension files
    const archive = archiver('zip');
     
    // listen for all archive data to be written
    // 'close' event is fired only when a file descriptor is involved
    output.on('close', () => {
      // console.log(archive.pointer() + ' total bytes');
      // console.log('archiver has been finalized and the output file descriptor has closed.');
      return resolve(true);
    });

    // This event is fired when the data source is drained no matter what was the data source.
    // It is not part of this library but rather from the NodeJS Stream API.
    // @see: https://nodejs.org/api/stream.html#stream_event_end
    output.on('end', () => {
      console.log('Data has been drained');
      return resolve(true);
    });
    
    // good practice to catch warnings (ie stat failures and other non-blocking errors)
    archive.on('warning', (err) => {
      if (err.code === 'ENOENT') {
        // log warning
        console.log(err);
      } else {
        // throw error
        return reject(err);
      }
    });
    
    // good practice to catch this error explicitly
    archive.on('error', (err) => {
      return reject(err);
    });
    
    // pipe archive data to the file
    archive.pipe(output);
  
    // append files from a sub-directory, putting its contents at the root of archive
    archive.directory(opts.extensionFilesDir, false);
  
    // finalize the archive (ie we are done appending files but streams have to finish yet)
    // 'close', 'end' or 'finish' may be fired right after calling this method so register to them beforehand
    archive.finalize();
  });
};

export default generateExtensionInfo;
