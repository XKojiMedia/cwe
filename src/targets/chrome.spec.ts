import { generateExtensionInfo, compileExtension } from './chrome';
import path from 'path';
import fs from 'fs';
import fsExtra from 'fs-extra';
import archiver from 'archiver';

jest.useFakeTimers();

jest.mock('path');
jest.mock('fs', () => {
  const _on = jest.fn((ev, cb) => {
    if (ev === 'end' || ev === 'close') {
      setTimeout(cb, 100000);
    }
  });
  return {
    createWriteStream: jest.fn(() => ({ on: _on })),
  };
});
jest.mock('fs-extra', () => {
  return {
    ensureFile: jest.fn(),
  }
});
const _archiveDirectory = jest.fn();
const _archive = {
    on: jest.fn(),
    pipe: jest.fn(),
    directory: _archiveDirectory,
    finalize: jest.fn(() => {
      jest.runAllTimers();
    }),
  };
jest.mock('archiver', () => {

  return jest.fn(() => _archive);
});

describe('chrome target', () => {
  describe('generateExtensionInfo', () => {
    it('should generate empty content if no manifest option', async () => {
      const extensionInfo = await generateExtensionInfo({
        exclude: [],
        include: [],
        outDir: '',
        rootDir: '',
        targets: [],
      });
      expect(extensionInfo.content).toBe('');
      expect(extensionInfo.fileName).toBe('manifest.json');
      expect(extensionInfo.fileType).toBe('json');
    });

    it('should generate the manifest content if manifest option provided', async () => {
      const extensionInfo = await generateExtensionInfo({
        exclude: [],
        include: [],
        outDir: '',
        rootDir: '',
        targets: [],
        manifestOptions: {
          name: 'Test extension',
          version: '0.0.1',
          background: {
            scripts: []
          },
          browser_action: {
            default_icon: '',
            default_title: '',
            default_popup: '',
          },
          content_security_policy: '',
          description: 'Test extension description',
          icons: '',
          offline_enabled: false,
          page_action: {
            default_icon: '',
            default_title: '',
            default_popup: '',
          },
          permissions: [],
          options_ui: {
            page: '',
            open_in_tab: false,
          },
          short_name: 'Test'
        }
      });
      expect(extensionInfo.content).toBeTruthy();
      expect(JSON.parse(extensionInfo.content)).toMatchSnapshot();
      expect(extensionInfo.fileName).toBe('manifest.json');
      expect(extensionInfo.fileType).toBe('json');
    });
  });

  describe('compileExtension', () => {
    it('should call archiver', async () => {
      await compileExtension({
        config: {
          exclude: [],
          include: [],
          outDir: '',
          rootDir: '',
          targets: [],
        },
        extensionBuildOutputDir: 'path/to/output',
        extensionFilesDir: 'path/to/files',
      });

      expect(archiver).toHaveBeenCalledWith('zip');
      expect(path.resolve).toHaveBeenCalledWith('path/to/output', 'chrome.zip');
      expect(_archiveDirectory).toHaveBeenCalledWith('path/to/files', false);
    });
    it('should throw an error if something fails', async () => {
      _archive.on.mockImplementation((ev, cb) => {
        if (ev === 'error') {
          throw new Error('fake!');
        }
      });
      await expect(compileExtension({
        config: {
          exclude: [],
          include: [],
          outDir: '',
          rootDir: '',
          targets: [],
        },
        extensionBuildOutputDir: 'path/to/output',
        extensionFilesDir: 'path/to/files',
      })).rejects.toThrowError();
    });
  });
});