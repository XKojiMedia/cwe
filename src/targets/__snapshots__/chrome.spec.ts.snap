// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`chrome target generateExtensionInfo should generate the manifest content if manifest option provided 1`] = `
Object {
  "background": Object {
    "scripts": Array [],
  },
  "browser_action": Object {
    "default_icon": "",
  },
  "content_security_policy": "",
  "description": "Test extension description",
  "icons": "",
  "manifest_version": 2,
  "name": "Test extension",
  "offline_enabled": false,
  "options_ui": Object {
    "open_in_tab": false,
    "page": "",
  },
  "permissions": Array [],
  "short_name": "Test",
  "version": "0.0.1",
}
`;
